Introduce intrusive split-list and compare its scanning performance with some other collections

xemul opened the pull request:
The purpose of this PR is to demonstrate how different "queues" (see below) perform on full-scan scenario.

Here a "queue" is a collection of objects that
* has std::deque-like ability to push_back and push_front elements to it
* can be scanned from start to end
* objects can be manipulated independently, in particular be of different types inheriting from some base class

Initially such a "queue" was met in CPU scheduler run-queue. In it, the stored element is `task*` and actual elements are classes that inherit from class task. The run-queue is implemented as `circular_buffer<task*>` container.

Scan performance comparison on 10k elements of different collections is:

On AMD Ryzen Threadripper 2970WX processor

GCC-15.2.1
|test|iters|runtime|±|allocs|tasks|inst|cycles|
|-|-|-|-|-|-|-|-|
|`vector<T>`|815590000|0.96ns|± 0.56%|0.000|0.000|7.04|2.9|
|`vector<T*>`|655960000|1.19ns|± 0.09%|0.000|0.000|7.04|3.8|
|`chunked_fifo<T*, 128>`|573820000|1.40ns|± 0.26%|0.000|0.000|8.08|4.6|
|`split_list<T, 16>`|504530000|1.55ns|± 1.45%|0.000|0.000|10.04|5.2|
|`intrusive_list<T>`|147000000|6.48ns|± 0.07%|0.000|0.000|5.04|25.8|

Clang-20.1.8
|test|iters|runtime|±|allocs|tasks|inst|cycles|
|-|-|-|-|-|-|-|-|
|`vector<T>`|1e+09|0.83ns|± 6.96%|0.000|0.000|7.00|3.2|
|`vector<T*>`|1e+09|0.93ns|± 0.25%|0.000|0.000|7.00|3.7|
|`split_list<T, 16>`|808970000|1.24ns|± 0.54%|0.000|0.000|8.50|4.9|
|`chunked_fifo<T*, 128>`|576560000|1.73ns|± 0.26%|0.000|0.000|13.04|6.9|
|`intrusive_list<T>`|153940000|6.20ns|± 0.75%|0.000|0.000|5.00|24.7|

On Intel i7-10750H processor
GCC-13.3.0
|test|iters|runtime|±|allocs|tasks|inst|cycles|
|-|-|-|-|-|-|-|-|
|`vector<T>`|920610000|0.95ns|± 0.19%|0.000|0.000|7.00|4.2|
|`vector<T*>`|832120000|1.18ns|± 0.33%|0.000|0.000|7.00|5.3|
|`split_list<T, 16>`|845490000|1.26ns|± 1.74%|0.000|0.000|10.00|5.4|
|`chunked_fifo<T*, 128>`|784010000|1.26ns|± 0.37%|0.000|0.000|8.04|5.8|
|`intrusive_list<T>`|124030000|8.16ns|± 1.11%|0.000|0.000|5.00|38.1|

On Intel(R) Xeon(R) Platinum 8559C processor (i7i.2xlarge instance)
Clang-20.1.2
|test|iters|runtime|±|allocs|tasks|inst|cycles|
|-|-|-|-|-|-|-|-|
|`vector<T>`|2e+09|0.43ns|± 0.01%|0.000|0.000|7.00|1.4|
|`vector<T*>`|2e+09|0.53ns|± 0.06%|0.000|0.000|7.00 |1.8|
|`split_list<T, 16>`|2e+09|0.65ns|± 0.19%|0.000|0.000|8.50|2.1|
|`chunked_fifo<T*, 128>`|1e+09|0.72ns|± 0.37%|0.000|0.000|13.04|2.4|
|`intrusive_list<T>`|188810000|5.33ns|± 0.10%|0.000|0.000|5.00|17.5|

dotnwat commented on tests/perf/container_perf.cc:
```diff
@@ -233,3 +238,239 @@ PERF_TEST_F(container_perf, iter_small_boost_deque) {
+        bi::cache_last<true>,
+        bi::member_hook<element, bi::slist_member_hook<>, &element::l_next>>;
+    static_assert(sizeof(slist) <= 2 * sizeof(void*));
+    slist _singly_linked_list;
```
curious if the intrusive list was what you reached for first but found the performance overhead to be too much? we had encountered this is doubly-linked intrusive list (even with all the safe/auto stuff turned off), and hand coded a simple version of effectively the same thing and got like a 10x performance reduction. not sure if slist has the same issues.

nyh commented:
This is a really interesting and clever implementation. I think its benefit over regular intrusive singly-linked lists is closely tied to how prefetching works on contemporary CPUs - that it actually prefetches data pointed by any pointer it sees, so it is helpful to let it see a few (F) pointers in advance instead of just the next one. I wonder how much this advantage remains if the scanning code does something more than just advance but actually do a bit more work on each item - in this case the pre-fetching of F items in advance might be too early - as we still have a lot of work to do before we're ready to need the Fth next item. Also in this case the speedup of just the scan might be negligible compared to the real amount of work we need to do on the items.

Do you know if something like your "split list" was ever done before? I searched a bit and got to a concept called a "chain" for each of the individual linked list, and "multi-chain" for the idea of having multiple independent lists each allowing one prefetching - but most discussions I read about these ideas where how to write a better prefetcher - not how to write a linked list that works better with the existing prefetcher. Maybe you can write a paper - or at least a blog post - on this new (?) technique :-)

You mentioned that you were looking for the ability to "push_back and push_front" - does your intended use case (run-queue) actually use both? chunked_fifo<> doesn't do push_front, but this was a decision to simplify the API - it's not that it was algorithmically impossible to implement. Of course if your goal is to be an _intrusive_ collection, then obviously chunked_fifo is not what you're looking for and you needed to create a new type. But in that case I think the word **intrusive** should be part of the new collection's name, and part of the pull request's title. If, on the other hand, "intrusive" was not very important to you - it appears that the existing chunked_fifo is actually a tiny bit more efficient than the new split_list, right?

nyh commented on include/seastar/util/split-list.hh:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    T* pop_front() noexcept {
+        auto i = _head_idx++ % F;
```
If this has a noticeable impact on performance, I think a comment recommending to use a power-of-two for F is in order. If the impact is very noticable, perhaps it makes sense to just force it, with a static_assert?

xemul commented on include/seastar/util/split-list.hh:
```diff
@@ -0,0 +1,120 @@
+
+/*
+ * Copyright (C) 2025 ScyllaDB
+ */
```
Yes, absolutely. The "add documentation" item is in the [TODO list](https://github.com/scylladb/seastar/pull/3138#issuecomment-3625810088)

> By the way, any reason why the older chunked_fifo.hh should be in core/ but this new one is in util/?

No specific reason, I just thought it was an auxiliary container, not "core" one whatever it means

nyh replied:
I think chunked_fifo.hh and circular_buffer.hh predate the separate directory for all the include files (include/seastar/), so they were in core/ because they were **used** in the core's implementation - and not really intended to be used by end-users. The placement in include/core is not because the header file is any more "core" or important for end-users than other header files.

In that sense, the new list is also used in the core's implementation so could be in core/ instead of util/ - like its other friends, chunked_fifo and circular_buffer.

But maybe this discussion is actually a sign that when created a new include/ directory, we shouldn't have kept the distinct "core" vs "util" subdirectory :-(

xemul commented:
upd:
- renamed the collection to be `intrusive_split_list`
- re-implemented properly the `operator->` and `operator*` for iterator
- initialize objects in base class constructor. This allows avoiding the ugly `init_once()` call and drop the manual start_measuring/stop_measuring tuning
- introduced generic tests parametrization. This allows having a single scanning test function implemented and change the collection size from the command-line

xemul commented:
Documentation is yet to be written :confused:

xemul commented:
> Looks like chunked_fifo wins

Looks like it doesn't :) I measured on i7i.2xlarge instance with Xeon processor -- split-list 2.1 vs chunked-fifo 2.4

xemul commented:
I updated the PR description with full table

xemul commented:
upd:
- added `size_t intrusive_split_list::size()` needed to implement task queue using this collection

nyh commented on include/seastar/util/split-list.hh:
```diff
@@ -0,0 +1,128 @@
+
+    public:
+        iterator(unsigned idx) noexcept
+                : _cur({nullptr})
```
By the way, I think the {nullptr} here is confusing on what it really does -
I think (but correct me if I'm wrong) that it only initializes the first entry to nullptr explicitly, and then by default initializes the other ones to nullptr as well. So I think just {} would do exactly the same thing.

mitso23 commented on tests/unit/split_list_test.cc:
```diff
@@ -0,0 +1,89 @@
+// checks that forward scanning reveals consistent results
+BOOST_AUTO_TEST_CASE(test_split_list_grow_and_shrink) {
+    const auto seed = std::random_device()();
+    auto rnd_engine = std::mt19937(seed);
```
You should also test iterator logic and maybe wrap around logic if you want to use 32bit unsigned integers for head/tail.

xemul commented on include/seastar/util/split-list.hh:
```diff
@@ -0,0 +1,128 @@
+template <typename T, unsigned F, T* T::*Member>
+class intrusive_split_list {
+    std::array<T*, F> _first;
+    std::array<T**, F> _last_p;
```
`T` can be an abstract type, like `class task`

mitso23 replied:
I see

xemul commented on include/seastar/util/split-list.hh:
```diff
@@ -0,0 +1,128 @@
+class intrusive_split_list {
+    std::array<T*, F> _first;
+    std::array<T**, F> _last_p;
+    unsigned _tail_idx;
```
Head wraps anyway (immediately after push_front() that happens before pop_front()). And 32 bits is more than enough for typical use-cases, seastar shard may have at most 64Gb, so 32-bit index is enough for 16-bytes objects, I don't think it's a problem
